### PR TITLE
Fix issues with llvm build and update scripts + documentation

### PR DESCRIPTION
**Pull request by StanSofomoDev**
fixes requested by @Jeff Nye: 

add a conditional source of sha.config or eliminate it. in build_llvm.sh
if [ -f "./sha.config" ]; then
 source ./sha.config
fi

change the build_llvm.sh script to specify the actual default source directory <something>/condor/riscv-llvm

change the update script to also suggest defaults.

have the update the script report the installation location

+ documentation changes

### CODE REVIEW

File: llvm/LLVM_FOR_RISCV.md [hunk 1]
```diff
@@ -109,19 +119,41 @@ Steps presented below expect that you have already build the LLVM on Linux for R
...
 
 - `source_dir`: Path to your LLVM source directory.
 - `build_dir`: Path to your LLVM build directory.
+
+You can also use the script with your own fusion definitions by using `-f [fusion_definitions]` argument:
+
 - `fusion_definitions`: Path to the file containing predefined fusion predicates (already included in the `how-to/llvm/macro_fusion` directory).
 
-Running this script will update the necessary LLVM source files with the predefined fusion definitions and automatically rebuild and install the updated LLVM components. After it succeeds you can run LLVM tools with `-mcpu='help'` argument to verify if new processor target with predefined fusions enabled is visible on the list.
+### Compile code with fusions enabled
```
Comment (jeffnye-gh) [inline llvm/LLVM_FOR_RISCV.md:148]: fusions is not plural

File: llvm/LLVM_INVESTIGATION_SUMMARY.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# Macro Fusion in LLVM for RISC-V
+
+Macro fusion is a performance optimization where specific instruction pairs are combined for more efficient execution, particularly useful in the RISC-V architecture. The LLVM compiler can be customized to support macro fusion through the modification of relevant target description files and by defining fusion rules in the `RISCVMacroFusion.td` file.
+
```
Comment (jeffnye-gh) [inline llvm/LLVM_INVESTIGATION_SUMMARY.md:4]: "particularly useful in the RISC-V architecture."

this is an opinion not suitable for a description of facts.

File: llvm/LLVM_INVESTIGATION_SUMMARY.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+For detailed instructions on configuring and enabling macro fusion in LLVM, please refer to the `LLVM_FOR_RISCV.md` file located in the **how-to** repository.
+
+## How Fusions Defined in `.td` Files Get Applied
+
+Fusions defined in `.td` files, like `RISCVMacroFusion.td`, are used to optimize instruction scheduling in the LLVM compilation process. These fusions are integrated through the auto-generated `RISCVGenMacroFusion.inc` and `RISCVGenSubtargetInfo.inc` files, which are built using the `TableGen` tool via the `RISCVCommonTableGen` target. This file contains important functionality for handling various processor-specific features, including macro fusion.
```
Comment (jeffnye-gh) [inline llvm/LLVM_INVESTIGATION_SUMMARY.md:9]: Fusions is not correct any more than informations is correct.

Use fusion groups or fusion tuples.

File: llvm/LLVM_INVESTIGATION_SUMMARY.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Additionally, `RISCVGenSubtargetInfo.inc` contains processor-specific definitions, which allow LLVM to handle various aspects of the scheduling process. These processor definitions include configurations for resources, buffer sizes, issue width, and scheduling models.
+
+## Multiple Instructions Fusions
+
+In LLVM's current implementation, RISC-V relies on the `SimpleFusion` class defined in `TargetMacroFusion.td` to enable macro fusion for two instructions at a time. This design hinges on the concept of "fusion targets"—specifically, first and second instruction targets—evaluated through predicates defined using `FusionPredicate`. These predicates check conditions such as opcode matching, operand values, and register usage, ensuring the fusion process is both valid and beneficial.
```
Comment (jeffnye-gh) [inline llvm/LLVM_INVESTIGATION_SUMMARY.md:19]: mention the LLVM version where you observed this.

**Review by jeffnye-gh — CHANGES_REQUESTED** [review]

File: llvm/LLVM_FOR_RISCV.md [hunk 5]
```diff
@@ -95,6 +95,16 @@ Compile the code:
...
+
+```bash
+[PATH_TO_YOUR_LLVM_INSTALL]/bin/clang -march=rv64gc -mabi=lp64d -save-stats hello.c -o hello
+```
+
+This will produce `hello.stats` file with llvm statistics for your compilation.
+
 ## Fusion Exploration with LLVM Compiler
 
 Macro fusion is a hardware optimization technique where certain pairs of instructions are fused together to execute more efficiently. This can lead to performance improvements without changing the existing Instruction Set Architecture (ISA). By defining preferred instruction pairs and hinting the compiler, we can increase the effectiveness of internal fusion hardware.
```
[the remark this answers is not in the document]
Comment (StanSofomoDev) [reply llvm/LLVM_FOR_RISCV.md:110]: Same here -> I think LLVM refers to `fusion tuples` as `predicators`

image

File: llvm/LLVM_INVESTIGATION_SUMMARY.md (hunk 3 above)
Comment (StanSofomoDev) [reply llvm/LLVM_INVESTIGATION_SUMMARY.md]: Done

File: llvm/LLVM_INVESTIGATION_SUMMARY.md (hunk 2 above)
Comment (StanSofomoDev) [reply llvm/LLVM_INVESTIGATION_SUMMARY.md:4]: Done

File: llvm/LLVM_FOR_RISCV.md (hunk 1 above)
Comment (StanSofomoDev) [reply llvm/LLVM_FOR_RISCV.md]: Done

File: llvm/LLVM_FOR_RISCV.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (StanSofomoDev) [reply llvm/LLVM_FOR_RISCV.md]: Done

**Pull request merged**